docs/scenarios/SINGLE_SIGN_ON_FOR_BOTS.md

```diff
@@ -0,0 +1,205 @@
+# Single Sign-On For Bots
+
+> ❗If you're using Azure Bot Service, please ensure that you've met the following prerequisites:
+> ### Using .NET
+> - Microsoft.Bot.Builder.Dialogs **v4.17.2** at minimum
+> - Microsoft.Bot.Builder.Integration.AspNet.Core **v4.17.2** at minimum
+
+## Using Bot Framework Web Chat Control
+
+```js
+const shareObservable = (observable) => {
+    let observers = [];
+    let subscription;
+
+    return new Observable(observer => {
+        if (!subscription) {
+            subscription = observable.subscribe({
+                complete() {
+                    observers.forEach(observer => observer.complete());
+                },
+                error(err) {
+                    observers.forEach(observer => observer.error(err));
+                },
+                next(value) {
+                    observers.forEach(observer => observer.next(value));
+                }
+            });
+        }
+
+        observers.push(observer);
+
+        return () => {
+            observers = observers.filter(o => o !== observer);
+            if (!observers.length) {
+                subscription.unsubscribe();
+                subscription = null;
+            }
+        };
+    });
+};
+
+class DefaultSubscriber {
+    applicable(activity) {
+        return true;
+    }
+
+    async next(activity) {
+        this.observer.next(activity);
+        return false;
+    }
+}
+
+const supportedSignInCardContentTypes = ["application/vnd.microsoft.card.signin", "application/vnd.microsoft.card.oauth"];
+const botOauthUrlRegex = /[\S]+.botframework.com\/api\/oauth\/signin\?signin=([\S]+)/;
+const signInIds = [];
+class BotAuthSubscriber {
+    applicable(activity) {
+        return activity && activity.attachments && activity.attachments.length && activity.attachments[0] && supportedSignInCardContentTypes.indexOf(activity.attachments[0].contentType) >= 0;
+    }
+
+    async apply(activity) {
+        this.observer.next(false); // Initially hides sign-in card
+
+        const attachment = activity.attachments[0];
+        const signInUrl = attachment.content.buttons[0].value;
+
+        // Extracts Sign In Id
+        const result = botOauthUrlRegex.exec(signInUrl);
+        const signInId = result[1];
+
+        if (!signInId) {
+            return;
+        }
+
+        // Ignore authenticated sign-in cards
+        if (signInIds.includes(signInId)) {
+            return;
+        }
+
+        // Extracts Bot Token URL
+        let botTokenUrl = undefined;
+        if (attachment && attachment.content && attachment.content.tokenPostResource && attachment.content.tokenPostResource.sasUrl) {
+            botTokenUrl = attachment.content.tokenPostResource.sasUrl;
+        }
+
+        if (!botTokenUrl) {
+            return;
+        }

```

Comment:
this seems to be candidate to be extracted to its own function, IMO will improve readability